org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java

```diff
@@ -621,37 +612,31 @@ boolean incorporate(InferenceContext18 context, TypeBound [] first, TypeBound []
 							// record all null tagBits as hints for the final inference solution.
 							long nullHints = (newConstraint.left.tagBits | newConstraint.right.tagBits) & TagBits.AnnotationNullMASK;
 							if (nullHints != 0) {
-								if (TypeBinding.equalsEquals(boundI.left, boundJ.left)
-										|| (boundI.relation == ReductionResult.SAME	&& TypeBinding.equalsEquals(boundI.right, boundJ.left))
-										|| (boundJ.relation == ReductionResult.SAME	&& TypeBinding.equalsEquals(boundI.left, boundJ.right))) {
-									boundI.nullHints |= nullHints;
-									boundJ.nullHints |= nullHints;
+								if (TypeBinding.equalsEquals(bound1.left, bound2.left)
+										|| (bound1.relation == ReductionResult.SAME	&& TypeBinding.equalsEquals(bound1.right, bound2.left))
+										|| (bound2.relation == ReductionResult.SAME	&& TypeBinding.equalsEquals(bound1.left, bound2.right))) {
+									bound1.nullHints |= nullHints;
+									bound2.nullHints |= nullHints;
 								}
 							}
 						}
 					}
-					ConstraintFormula[] typeArgumentConstraints = deriveTypeArgumentConstraints ? deriveTypeArgumentConstraints(boundI, boundJ) : null;
+					ConstraintFormula[] typeArgumentConstraints = deriveTypeArgumentConstraints ? deriveTypeArgumentConstraints(bound1, bound2) : null;
 					if (typeArgumentConstraints != null) {
-						for (int k = 0, length = typeArgumentConstraints.length; k < length; k++) {
-							if (!reduceOneConstraint(context, typeArgumentConstraints[k]))
+						for (ConstraintFormula typeArgumentConstraint : typeArgumentConstraints) {
+							if (!reduceOneConstraint(context, typeArgumentConstraint))
 								return false;
 						}
 					}
 					if (iteration == 2) {
-						TypeBound boundX = boundI;
-						boundI = boundJ;
-						boundJ = boundX;
+						TypeBound boundX = bound1;
+						bound1 = bound2;
+						bound2 = boundX;
 					}
 				} while (first != next && ++iteration <= 2);
 			}
 		}
-		/* TODO: are we sure this will always terminate? Cf. e.g. (Discussion in 18.3):
-		 *
-		 *    "The assertion that incorporation reaches a fixed point oversimplifies the matter slightly. ..."
-		 */
```

Comment:
someone is stealing comments.

Reply:
Someone called "cleanup Java 5: Convert to enhanced 'for' loops". Nice finding. Consider opening a bug on jdt.ui